query-engine/connectors/sql-query-connector/src/filter/visitor.rs

```diff
@@ -213,260 +305,283 @@ impl AliasedCondition for ScalarFilter {
+        ctx: &Context<'_>,
+    ) -> (ConditionTree<'static>, Option<Vec<AliasedJoin>>) {
+        let parent_alias = self.parent_alias();
+        let parent_alias_string = parent_alias.as_ref().map(|a| a.to_string(None));
+
+        // If the relation is inlined, we simply check whether the linking fields are null.
+        //
+        // ```sql
+        //  SELECT "Parent"."id" FROM "Parent"
+        //    WHERE "Parent"."childId" IS NULL;
+        // ```
+        if filter.field.is_inlined_on_enclosing_model() {
+            let conditions: Vec<_> = ModelProjection::from(filter.field.linking_fields())
+                .as_columns(ctx)
+                .map(|c| c.opt_table(parent_alias_string.clone()))
+                .map(|c| c.is_null())
+                .map(Expression::from)
                 .collect();
 
-            let join_columns: Vec<Column> = self
-                .field
-                .join_columns(ctx)
-                .map(|c| c.aliased_col(Some(alias), ctx))
-                .collect();
+            return (ConditionTree::And(conditions), None);
+        }
 
-            let related_table = self.field.related_model().as_table(ctx);
-            let related_join_columns: Vec<_> = ModelProjection::from(self.field.related_field().linking_fields())
+        // If the relation is not inlined and we can use joins, then we join the relation and check whether the related linking fields are null.
+        //
+        // ```sql
+        //  SELECT "Parent"."id" FROM "Parent"
+        //    LEFT JOIN "Child" AS "j1" ON ("j1"."parentId" = "Parent"."id")
+        //  WHERE "j1"."parentId" IS NULL OFFSET;
+        // ```
+        if self.should_render_join() {
+            let alias = self.next_alias(AliasMode::Join);
+
+            let conditions: Vec<_> = ModelProjection::from(filter.field.related_field().linking_fields())
                 .as_columns(ctx)
-                .map(|col| col.aliased_col(Some(alias.flip(AliasMode::Join)), ctx))
+                .map(|c| c.aliased_col(Some(alias), ctx))
+                .map(|c| c.is_null())
+                .map(Expression::from)
                 .collect();
 
-            let nested_conditions = self
-                .nested_filter
-                .aliased_condition_from(Some(alias.flip(AliasMode::Join)), false, ctx)
-                .invert_if(condition.invert_of_subselect());
-
-            let conditions = selected_identifier
-                .clone()
-                .into_iter()
-                .fold(nested_conditions, |acc, column| acc.and(column.is_not_null()));
-
-            let join = related_table
-                .alias(alias.to_string(Some(AliasMode::Join)))
-                .on(Row::from(related_join_columns).equals(Row::from(join_columns)));
+            let join = compute_one2m_join(
+                &filter.field,
+                alias.to_string(None).as_str(),
+                parent_alias_string.as_deref(),
+                ctx,
+            );
```

Comment:
This part is the most important change. When traversing a `{ to_one: null }`, we're now using a join if we can.